ext_mongodb.php

```diff
@@ -378,65 +378,229 @@ public function __debugInfo()
 }
 
 final class Query {
-	private array $query;
+	private mixed $filter = NULL;
+	private array $opts = [];
+	private string $readConcernLevel = NULL;
 
-	public function __construct(mixed $filter, array $options = array())
+	/* Helpers for setting / validating scalar options */
+	private function _queryOptBool( string $optName, array $options, string $key )
 	{
-		$zquery = [];
+		if ( array_key_exists( $key, $options ) )
+		{
+			$this->opts[$optName] = (bool) $options[$key];
+			return true;
+		}
 
-		/* phongo_query_init */
-		Utils::mustBeArrayOrObject('parameter 1', $filter, "MongoDB\Driver\Query::__construct");
+		return false;
+	}
+	
+	private function _queryOptDocument( string $optName, array $options, string $key )
+	{
+		if ( array_key_exists( $key, $options ) )
+		{
+			if ( !is_array( $options[$key] ) && !is_object( $options[$key] ) )
+			{
+				throw new \MongoDB\Driver\Exception\InvalidArgumentException(
+					sprintf(
+						'Expected "%s" %s to be array or object, %s given',
+						$key, $key[0] == '$' ? 'modifier' : 'option', gettype( $options[$key] )
+					)
+				);
+			}
+			$this->opts[$optName] = (object) $options[$key];
+			return true;
+		}
 
-		if ($options) {
-			$this->query['batchSize'] = array_key_exists('batchSize', $options ) ? (int) $options['batchSize'] : 0;
-			$this->query['flags'] = array_key_exists('flags', $options ) ? (int) $options['flags'] : 0;
-			$this->query['limit'] = array_key_exists('limit', $options ) ? (int) $options['limit'] : 0;
-			$this->query['skip'] = array_key_exists('skip', $options ) ? (int) $options['skip'] : 0;
-
-			if (array_key_exists('readConcern', $options)) {
-				if (!($options['readConcern'] instanceof \MongoDB\Driver\readConcern)) {
-					throw new \MongoDB\Driver\Exception\InvalidArgumentException(
-						'Expected "readConcern" option to be MongoDB\Driver\ReadConcern, ' .
-						gettype($options['readConcern']) . ' given'
-					);
-				} else {
-					if ($options['readConcern']->getLevel() != NULL) {
-						$this->query['readConcern'] = $options['readConcern']->getLevel();
-					}
-				}
+		return false;
+	}
+
+	private function _queryOptInt64( string $optName, array $options, string $key )
```

Comment:
libmongoc seems to be strict about the find/cursor options being 64-bit integers (see: [`_mongoc_cursor_set_opt_int64()`](https://github.com/mongodb/mongo-c-driver/blob/1.5.0-rc0/src/mongoc/mongoc-cursor.c#L54)). I don't think this affects you, as you construct the options once and never use other helpers to change them. It also shouldn't be an issue reading the options, since [`_mongoc_cursor_get_opt_int64()`](https://github.com/mongodb/mongo-c-driver/blob/1.5.0-rc0/src/mongoc/mongoc-cursor.c#L74) will cast the value to an `int64_t` on the way out.

That said, it's good to be aware of this. If your BSON encoding is similar to PHPC, you elect to encode as an `int32_t` or `int64_t` based on the integer value.